File: ZombustersWindows/SteeringBehaviors/SteeringEntity.cs
using System;
using System.Collections.Generic;
using Microsoft.Xna.Framework;

namespace ZombustersWindows
{
    public class SteeringEntity
    {
        private Vector2 _velocity;
        private Vector2 _position;
        private Vector2 _previousPosition;
        private float _maxSpeed;
        private float _boundingRadius;
        private float _width;
        private float _height;

        /// <summary>
        /// Ángulo hacia el cual "mira" la entidad
        /// </summary>
        public float Angle
        {
            get
            {
                return VectorHelper.GetAngle(_previousPosition, _position);
            }
        }

        public float Width
        {
            get
            {
                return _width;
            }
            set
            {
                _width = value;
            }
        }

        public float Height
        {
            get
            {
                return _height;
            }
            set
            {
                _height = value;
            }
        }

        /// <summary>
        /// Radio del círculo bounding
        /// </summary>
        public float BoundingRadius
        {
            get { return _boundingRadius; }
            set { _boundingRadius = value; }
        }

        /// <summary>
        /// Vector velocidad
        /// </summary>
        public Vector2 Velocity
        {
            get { return _velocity; }
            set { _velocity = value; }
        }

        /// <summary>
        /// Orientación de la entidad
        /// </summary>
        public Vector2 Heading
        {
            get
            {
                return new Vector2(
                    (float)Math.Abs(Math.Cos(this.Angle)), (float)Math.Abs(Math.Sin(this.Angle)));
            }
        }

        /// <summary>
        /// Perpendicular al vector heading
        /// </summary>
        public Vector2 Side
        {
            get
            {
                return VectorHelper.GetPerpendicular(this.Heading);
            }
        }

        /// <summary>
        /// Velocidad
        /// </summary>
        public float Speed
        {
            get { return _velocity.Length(); }
        }

        /// <summary>
        /// Posición de la entidad en el "mundo 2d"
        /// </summary>
        public Vector2 Position
        {
            get { return _position; }
            set
            {
                _previousPosition = _position;
                _position = value;
            }
        }

        public Vector2 PreviousPosition
        {
            get { return _previousPosition; }
        }

        /// <summary>
        /// Velocidad máxima de la entidad
        /// </summary>
        public float MaxSpeed
        {
            get { return _maxSpeed; }
            set { _maxSpeed = value; }
        }

    }
}

File: shared/src/commonMain/kotlin/com/retrowax/zombusters/game/enemy/SteeringEntity.kt
package com.retrowax.zombusters.game.enemy

class SteeringEntity(
    initialPosition: Vec2,
    var maxSpeed: Float = 1.5f,
    var boundingRadius: Float = 5f,
    var width: Float = 0f,
    var height: Float = 0f
) {
    var velocity: Vec2 = Vec2.ZERO

    private var _position: Vec2 = initialPosition.copy()
    private var _previousPosition: Vec2 = initialPosition.copy()

    var position: Vec2
        get() = _position
        set(value) {
            _previousPosition = _position.copy()
            _position = value.copy()
        }

    val previousPosition: Vec2 get() = _previousPosition

    val speed: Float get() = velocity.length()

    // Heading derived from velocity direction; falls back to east (1,0) when stationary
    val heading: Vec2 get() {
        val len = velocity.length()
        return if (len > 0.001f) Vec2(velocity.x / len, velocity.y / len) else Vec2(1f, 0f)
    }

    // Perpendicular to heading (-heading.y, heading.x)
    val side: Vec2 get() { val h = heading; return Vec2(-h.y, h.x) }
}
